designs/2022-forbid-disable/README.md

```diff
@@ -0,0 +1,120 @@
+- Repo: eslint/eslint
+- Start Date: 2022-02-28
+- RFC PR:
+- Authors: Brian Bartels
+
+# ESLintRC Option To Forbid Disable For Specific Rules
+
+## Summary
+
+Allow users to forbid users from disabling certain rules. This is similar to `noInlineConfig` where the primary difference is that `noInlineConfig` is set for all rules.
+
+## Motivation
+
+When working with complex lint configs. Some rules are more important than other. For this reason it can be inconsequential to disable certain rules. But incredibly bad to disable others.
+
+EG:
+```
+//eslint-disable-next-line @government/required-SECURITY-rule
+```
+Is a much worse rule to disable than
+```
+//eslint-disable-next-line noExtraSemi
+```
+
+`noInlineConfig` works well for some. But in reality, many users disable rules at least somewhat frequently for valid reasons. And in large repos, there are going to be exceptions that need inline configuration. 
+
+## Detailed Design
+
+in `linter.js` I will adjust the `getDirectiveComments` function so that I reference a list of forbidden rules. If someone is disabling a rule that is forbidden. I will add an error similar to `warnInlineConfig`
+
+In `eslintrc` I will add another option for rules. The new state can be
+```
+"warn"
+"error"
+"forbid"
+```
+where forbid is always an error and cannot be overridden by a directive
+
+I will add a new array called `forbiddenRules` that will be processed on initialization of the linter.js. This will happen in the Constructor. Then in `getDirectiveComments` I will reference the rule being disabled in the directive. If that rule matches a rule in `forbiddenRules` then I will not accept that directive. And I will declare an error on that string
+
+EG `// eslint-disable semi` will have ~~semi~~ as an error
```

Comment:
What will be the behavior with directives that are disabling all rules? Those are directives that don't specify any particular rule, for example `/* eslint-disable-next-line */`

Reply:
hmm that is a feature that I didn't consider. When are those disable's calculated? I'm assuming that I'd need to modify `runRules` in some way to handle this.

One feature that I'm extra concerned about is
```
/* eslint-disable */
alert('foo');
/* eslint-enable */
```

Reply:
They are prepared as disable directive objects with `ruleId: null` (which means "all rules") in [`createDisableDirectives`](https://github.com/eslint/eslint/blob/802c85c28283362d48d77876e924fd381a15f573/lib/linter/linter.js#L297), and then applied in [`applyDisableDirectives`](https://github.com/eslint/eslint/blob/802c85c28283362d48d77876e924fd381a15f573/lib/linter/apply-disable-directives.js#L293) along with other disable directives.

Reply:
hmm okay I pushed a few changes. I'm feeling like my proposed design using `getDirectiveComments` isn't able to handle this. Would a different solution where I delay this processing entirely to `applyDisableDirectives` work better. Where I calculate the `forbid` rules for the file and filter them out there?

Reply:
Okay so this is fixed. But there will not be an error on the directive where `ruleId: null` since I can't easily match that to the list

Reply:
Isn't that a major loophole? Someone wanting to avoid this restriction could use `/* eslint-disable-next-line */` and PR reviewers likely wouldn't know they are trying to avoid the dangerous rule so it could easily slip through. This seems to me like it would negate much of the benefit of this new feature.

To solve this, we could couple this feature with another new feature based on [eslint-comments/no-unlimited-disable](https://mysticatea.github.io/eslint-plugin-eslint-comments/rules/no-unlimited-disable.html) which doesn't allow the "unlimited disables".

Reply:
> Isn't that a major loophole? Someone wanting to avoid this restriction could use `/* eslint-disable-next-line */`

I think the idea is that neither `/* eslint-disable-next-line semi */` nor `/* eslint-disable-next-line */` can disable `semi` problems.

The difference is that `/* eslint-disable-next-line semi */` itself will be reported as a lint problem (attempt to disable a rule that cannot be disabled) while `/* eslint-disable-next-line */` won't.

Reply:
I see, that makes sense.

One concern is that this makes comments like `/* eslint-disable-next-line */` less straightforward. A comment that appears to disable all rules now has exceptions and requires checking the global config to determine the actual behavior.

Reply:
> One concern is that this makes comments like `/* eslint-disable-next-line */` less straightforward. A comment that appears to disable all rules now has exceptions and requires checking the global config to determine the actual behavior.

Indeed, that could be confusing for users. Any ideas on how we can avoid this confusing behavior?

Reply:
If we're okay with including more information in the violation message, and if we think this situation warrants it, the subsequent violation message could explain why the preceding disable directive does not apply.

```
// eslint-disable-next-line
console.log(123);
```

```
Unexpected console statement eslint(no-console)
A disable directive for this violation was detected but `noInlineConfig` is enabled for this rule. 
```